Partial support for Translations

n76 opened the pull request:
Partial implemented support for the languages and dialets that Roku supports.

A number of structural issues encountered the biggest are:
1. Programatically catenating strings which assumes a particular language syntax.
2. Using display strings in buttons, etc. to determine action.
3. Building display objects in BrightScript rather than in XML. This results in many cases where you have to translate the same string many times (e.g. "Cancel")

Side note: My editor is set to remove trailing spaces and to assure a new line at the end of the file. Existing source code seems to vary on this. I can revert those few changes if that is a problem.

Finally, added https://reuse.software/spec/ compliant license information to the new files and the files I touched.

Signed-off-by: Tod Fitch <tod@fitchfamily.org>

<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://jellyfin.readthedocs.io/en/latest/developer-docs/contributing/ page.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

neilsb commented on components/tvshows/TVListDetails.brs:
```diff
@@ -17,7 +21,7 @@ function itemContentChanged() as void
   if type(itemData.RunTimeTicks) = "LongInteger"
     m.top.findNode("runtime").text = stri(getRuntime()).trim() + " mins"
-    m.top.findNode("endtime").text = "Ends at " + getEndTime()
+    m.top.findNode("endtime").text = tr("Ends at ") + getEndTime()
```
"Ends at" is not being translated when I'm testing locally with German.  Not sure if it's anything to do with the trailing space here and in the definitions.

cewert replied:
Either way we should probably get rid of the trailing spaces.  If someone make a new page using "Ends at" then it wouldn't get translated correct? Unless they also had a trailing space

n76 replied:
Found a that you can put [parameters in the translation strings](https://developer.roku.com/docs/developer-program/core-concepts/localization.md), so made that change.

neilsb commented:
That's a great start, and the structural issue you highlighted are things we should look at addressing before the client grows, which will just compound the problem.

The REUSE Specification is not something I've come across before.  I'm not sure if we want that at the top of every file, but certainly happy to have a discussion on it.  Might I suggest that we move that into it's own PR or issue?  And if we're going to do it we should probably hit all the files at the one time and then it becomes the standard.  On thing, though, is that I believe the Jellyfin project, and the roku client both use _GPL v2_, as opposed to _GPL v2 or later_, just looking at the current license file.

anthonylavado commented:
For licensing, yes, let's please move that in to a separate PR if we can?

I've edited the title to reflect what I believe was the intention of this PR.

It's 5 am here, so when I'm properly awake I'll look over the changes for translation. We use Weblate for the rest of the project, so if we can set up string files in a compatible/supported format, I can add this repository to the software, enabling better translation support.

neilsb commented:
One for the formats Roku supports is  XLIFF, which is what @n76 has used.  XLIFF is also conveniently supported by Weblate according to their docs.  So that's positive.

n76 commented:
I've removed the license info headers and fixed the 'Ends at ' issue. At least for movies. I am having issues that I believe are unrelated to my changes in viewing detailed data for TV shows and other collections.

There is another case similar to the 'Ends at ' for entering settings in components/config/ConfigList.brs at line 51: `dialog.title = "Enter the " + configField.label`. The problem being that the `configField.label` has already been translated. But simply fixing this the same as 'Ends at ' assumes the same word order, declensions, etc. as English. I think this will have to be refactored to be pass the full string in the object.

neilsb commented on components/tvshows/TVListDetails.brs:
```diff
@@ -17,7 +17,7 @@ function itemContentChanged() as void
   if type(itemData.RunTimeTicks) = "LongInteger"
     m.top.findNode("runtime").text = stri(getRuntime()).trim() + " mins"
-    m.top.findNode("endtime").text = "Ends at " + getEndTime()
+    m.top.findNode("ends-at", tr("Ends at %1").Replace("%1", getEndTime()))
```
This is the error with viewing the TV Show details.   Suspect it was a copy & paste error.  :)

n76 replied:
Yep. After this commit maybe I'll better learn the code base by trying to consolidate common logic into some functional utility routines. That might cut down the use of copy and  paste.

And this reminds me again how bad the Roku development process is: I'd really, really like to be able to do lexical, syntax and semantics check on everything before pushing it to the TV. And I'd really, really like if if the TV didn't crash and need power cycling when you have some specific syntax/semantic errors.

cewert commented:
@n76 curious what made you go with the XLIFF format?

It looks like we can choose either TS or XLIFF. I don't know all the differences between the two but I noticed XLIFF uses ID's and TS doesn't. So to me TS seems easier to maintain since we wouldn't have to worry about the ID but I assume there's other factors to consider as well.

n76 commented:
> @n76 curious what made you go with the XLIFF format?
> 
> It looks like we can choose either [TS or XLIFF](https://developer.roku.com/en-ca/docs/developer-program/core-concepts/localization.md#ts-file-example). I don't know all the differences between the two but I noticed XLIFF uses ID's and TS doesn't. So to me TS seems easier to maintain since we wouldn't have to worry about the ID but I assume there's other factors to consider as well.

I used XLIFF on this because I'd used it on my [PeerTube Roku channel](https://github.com/n76/PeerVue). Now I have to remember why I used it there instead of TS. And I think it was just a toss of the coin. I don't know how critical the ID numbers are in XLIFF as far as Roku is concerned and admit they are a bit of a hassle to keep up. If this is an issue, I'll look into converting the XLIFF to TS.

By the way, writing that PeerTube channel was inspired by seeing this project. And also what got me interested in the licensing issue as I borrowed the make files and a utility file from here but the licensing I used was different from GPL2 only so I need a way to keep the GPL2 only licensing on the borrowed files. That turned out useful when I needed a custom font (GPL3 or later) and some icons (Apache 2.0).

n76 commented:
Conversion from XLIFF to TS was pretty simple and seems to work the same. I know almost nothing of Weblate but it looks like it can deal with TS files as well as XLIFF files.

Should I update the pull request to use TS?

neilsb commented:
Seems that either should be fine with Webalate, but if using TS means we don't need to manually manage an id, then I'd say that would be preferable.

neilsb commented:
That's great.  One thing we'll need to be aware of now is when laying things out a label, title, button, etc may be a different size depending on the language.

There is only one example of this I found currently, and that's when you're using German. The "* Options" on the JFOverhang is now right up against the right hand edge of the screen.  I suspect this could be solved using a right aligned layout group or something.  Might be worth sorting this out before merging.

n76 commented:
> That's great. One thing we'll need to be aware of now is when laying things out a label, title, button, etc may be a different size depending on the language.
> 
> There is only one example of this I found currently, and that's when you're using German. The "* Options" on the JFOverhang is now right up against the right hand edge of the screen. I suspect this could be solved using a right aligned layout group or something. Might be worth sorting this out before merging.

The "* Options" positioning is complicated by the fact that it is two different text strings and the sizes are different (set different in the XML and then changed in the BRS). My initial searches to see if you can vary font size, etc. within a string have come up blank. Next search was to see if one element position can be made relative to another. First search on that has come up blank too. My next search was to see if there was a call to determine the actual display width of text (then the "*" could be repositioned to be just to the left of the "Options" text). No luck there either.

Not to say that any of those things can't be done as I've had trouble finding things in the Roku documents before. Just that the searches I made for the first few ways to address this that I thought of came up empty. It looks like Roku themselves know how to do it if you look at the time display, I just haven't found the way yet.

In the meantime, I'll push a change where I manually moved the "* Options" to the left by a few pixels. Making it look good in one language does upset the look a little for other languages.

neilsb commented on components/JFOverhang.xml:
```diff
@@ -35,9 +35,9 @@
-    
-    <Label id="overlayOptionsStar" font="font:LargeSystemFont" text="*" translation="[1791, 96]" />
-    <Label id="overlayOptionsText" font="font:SmallSystemFont" text="Options" translation="[1821, 103]" />
+
+    <Label id="overlayOptionsStar" font="font:LargeSystemFont" text="*" translation="[1780, 96]" />
+    <Label id="overlayOptionsText" font="font:SmallSystemFont" text="Options" translation="[1810, 103]" />
```
I see what you mean.  A bit of fiddling and I think the custom vertical alignment option may give us a neat way of doing it.
```suggestion
    <LayoutGroup layoutDirection="horiz" horizAlignment="right" translation="[1900, 96]" vertAlignment="custom">
      <Label id="overlayOptionsStar" font="font:LargeSystemFont" text="*"   />
      <Label id="overlayOptionsText" font="font:SmallSystemFont" text="Options" translation="[0,6]" />
    </LayoutGroup>
```

n76 replied:
That does seem to work. I've changed the pull request to use it.
